File: Roadmap/02 - FUNCIONES Y ALCANCE/php/gabrielmoris.php
<?php

/*
*EXERCISE:
* - Create examples of basic functions that represent the different
* possibilities of the language:
* - Without parameters or return, with one or more parameters, with return...
* - Check if you can create functions within functions.
* - Use some examples of functions already created in the language.
* - Test the concept of LOCAL and GLOBAL variables.
* - You must print the result of all the examples to the console.
* (and keep in mind that each language may have more or fewer possibilities)
* EXTRA DIFFICULTY (optional):
* Create a function that receives two string parameters and returns a number.
* - The function prints all the numbers from 1 to 100. Considering that:
* - If the number is a multiple of 3, it displays the string of the first parameter.
* - If the number is a multiple of 5, it displays the string of the second parameter.
* - If the number is a multiple of 3 and 5, it displays the two concatenated strings.
* - The function returns the number of times the number has been printed instead of the texts.
* Pay special attention to the syntax you must use in each case.
* Each language follows conventions that you must respect for the code to be understood.
*/

// simple function

function simplicity()
{
    echo "ok";
};

// One param without return function
function one_param($param)
{
    echo  "This is the param: " . $param . "\n";
};

one_param("hi!");

// 2 Params with return
function two_params_Return($first, $second)
{
    return $first + $second;
}

$asign_to_var = two_params_Return(1, 5);
echo $asign_to_var . "\n";

// Functions in functions

function outside($res)
{
    function inside($first, $second)
    {
        return $first + $second;
    };

    return inside($res, 1);
};

echo outside(5) . "\n";

// Recursive

$i = 0;

function make_i_great_again()
{
    global  $i; // in this way I access it as a global from outside. Otherwise it behaves local.
    if ($i > 99) {
        return $i;
    };
    $i++;
    return make_i_great_again($i);
}

echo make_i_great_again() . "\n";
// native methods

$hardcodedArr = [10, 11, 12];

function sum($carry, $item)
{
    return $carry + $item;
}

$result = array_reduce($hardcodedArr, 'sum', 0);
echo $result . "\n";

// Global variable: Outside any function or class
$globalVar = "This is a global variable";

// Local variable: Inside a function or class
function myFunction()
{
    $localVar = "This is a local variable";
    echo $localVar; // Output: "This is a local variable"
}
// echo $localVar; // error

function flix_bus($str, $str2)
{
    $num_count = 0;
    for ($i = 0; $i <= 100; $i++) {
        if ($i % 3 == 0 && $i % 5 == 0) {
            echo $str . "," . $str2;
        } elseif ($i % 3 == 0) {
            echo $str . ",";
        } elseif ($i % 5 == 0) {
            echo $str2 . ",";
        } else {
            echo $i . ",";
            $num_count++;
        }
    }

    return $num_count;
}

$result = flix_bus("Flix", "Bus");

echo "\n Result: " . $result . "\n";

File: Roadmap/03 - ESTRUCTURAS DE DATOS/go/gabrielmoris.go
/*
 * EXERCISE:
 * - Create examples of basic functions that represent the different
 *   possibilities of the language:
 *   Without parameters or return, with one or more parameters, with return...
 * - Check if you can create functions within functions.
 * - Use some examples of functions already created in the language.
 * - Test the concept of LOCAL and GLOBAL variables.
 * - You must print the result of all examples to the console.
 *   (and keep in mind that each language may have more or fewer possibilities)
 *
 * EXTRA DIFFICULTY (optional):
 * Create a function that receives two parameters of string type and returns a number.
 * - The function prints all numbers from 1 to 100. Taking into account that:
 *   - If the number is a multiple of 3, it shows the text string of the first parameter.
 *   - If the number is a multiple of 5, it shows the text string of the second parameter.
 *   - If the number is a multiple of 3 and 5, it shows both text strings concatenated.
 *   - The function returns the number of times the number has been printed instead of the texts.
 *
 * Pay special attention to the syntax you should use in each of the cases.
 * Each language follows conventions that you must respect for the code to be understood.
 */

package main

import "fmt"
 var globalvar = "I am global because I am outside any function or block."
 func main() {
	var localVar ="I am local because I am inside a function";
	fmt.Println(localVar);
	fmt.Println(noParams()); // No params
	fmt.Println(withParams(1,2)) // 3
	var number = 1;
	notReturn(&number);
	fmt.Println(number);  // 2
	b(a); // string
	recursive(&number,10);

	// Anonymous Functions
	anonFunc := func() {
        fmt.Println("I am an anonymous function")
    }
	anonFunc();

	// Clousures
	counter := 0
    increment := func() int {
        counter++
        return counter
    }
    fmt.Println(increment()) // 1
    fmt.Println(increment()) // 2
	fmt.Println(sum(1,2,3,4,5)) //15
	challenge("fizz","buzz");
 }

 // Without Params
 func noParams() string {
	return "No Params";
 }

 // With Params
 func withParams(num1 int, num2 int) int{
	return num1 + num2;
 }

 // Function modifying external variable
 func notReturn(num *int){
	*num += 1;
 }

 // High order Functions
 func a () string {
	return "I am func A"
 }

 func b (function func() string) {
	fmt.Println("I am in func B and I an calling this fn: " + function());
 }

 // Recursive Function
 func recursive (num *int , max int){
	if *num <= max {
		fmt.Println("This is recutsive: ")
		fmt.Println(*num)
		*num++
		recursive(num, max)
	}
 }

 // Variable number of arguments
 func sum(nums ...int) int {
    total := 0
    for _, num := range nums {
        total += num
    }
    return total
}

func challenge(str1 string, str2 string){
	for i:=0; i<100; i++{
		if(i %3 ==0 && i%5==0){
			fmt.Println(str1 +  " "+ str2)
		} else if(i % 3==0){
			fmt.Println(str1)
		} else if(i % 5==0){
			fmt.Println(str2)
		}

	}

}